From 7d20a418593bba8a21cfdf9fbfdb40735c291710 Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Date: Sat, 27 Jun 2020 01:47:28 +0300
Subject: [PATCH 3/5] libcamera: Add support for MediaTek packed Bayer formats

The MediaTek MT8183 camera hardware stores 10- and 12-bit Bayer data in
a vendor-specific packed format. Support it in libcamera.

The 14-bit version is left out for the time being as its alignment
constraints are not clear yet.

Upstream-Status: Pending

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
(cherry picked from commit 2e43179a3a84ddd4bdc36c921ea79a3c793b7217)
Signed-off-by: Vitor Sato Eschholz <vsatoes@baylibre.com>
---
 src/libcamera/formats.cpp          | 80 ++++++++++++++++++++++++++++++
 src/libcamera/formats.yaml         | 26 ++++++++++
 src/libcamera/v4l2_pixelformat.cpp | 16 ++++++
 3 files changed, 122 insertions(+)

diff --git a/src/libcamera/formats.cpp b/src/libcamera/formats.cpp
index 447e6238..7fffdf28 100644
--- a/src/libcamera/formats.cpp
+++ b/src/libcamera/formats.cpp
@@ -880,6 +880,86 @@ const std::map<PixelFormat, PixelFormatInfo> pixelFormatInfo{
 		.pixelsPerGroup = 25,
 		.planes = {{ { 32, 1 }, { 0, 0 }, { 0, 0 } }},
 	} },
+	{ formats::SBGGR10_MTK, {
+		.name = "SBGGR10_MTK",
+		.format = formats::SBGGR10_MTK,
+		.v4l2Formats = { V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SBGGR10), },
+		.bitsPerPixel = 10,
+		.colourEncoding = PixelFormatInfo::ColourEncodingRAW,
+		.packed = true,
+		.pixelsPerGroup = 4,
+		.planes = {{ { 5, 1 }, { 0, 0 }, { 0, 0 } }},
+	} },
+	{ formats::SGBRG10_MTK, {
+		.name = "SGBRG10_MTK",
+		.format = formats::SGBRG10_MTK,
+		.v4l2Formats = { V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SGBRG10), },
+		.bitsPerPixel = 10,
+		.colourEncoding = PixelFormatInfo::ColourEncodingRAW,
+		.packed = true,
+		.pixelsPerGroup = 4,
+		.planes = {{ { 5, 1 }, { 0, 0 }, { 0, 0 } }},
+	} },
+	{ formats::SGRBG10_MTK, {
+		.name = "SGRBG10_MTK",
+		.format = formats::SGRBG10_MTK,
+		.v4l2Formats = { V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SGRBG10), },
+		.bitsPerPixel = 10,
+		.colourEncoding = PixelFormatInfo::ColourEncodingRAW,
+		.packed = true,
+		.pixelsPerGroup = 4,
+		.planes = {{ { 5, 1 }, { 0, 0 }, { 0, 0 } }},
+	} },
+	{ formats::SRGGB10_MTK, {
+		.name = "SRGGB10_MTK",
+		.format = formats::SRGGB10_MTK,
+		.v4l2Formats = { V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SRGGB10), },
+		.bitsPerPixel = 10,
+		.colourEncoding = PixelFormatInfo::ColourEncodingRAW,
+		.packed = true,
+		.pixelsPerGroup = 4,
+		.planes = {{ { 5, 1 }, { 0, 0 }, { 0, 0 } }},
+	} },
+	{ formats::SBGGR12_MTK, {
+		.name = "SBGGR12_MTK",
+		.format = formats::SBGGR12_MTK,
+		.v4l2Formats = { V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SBGGR12), },
+		.bitsPerPixel = 12,
+		.colourEncoding = PixelFormatInfo::ColourEncodingRAW,
+		.packed = true,
+		.pixelsPerGroup = 2,
+		.planes = {{ { 3, 1 }, { 0, 0 }, { 0, 0 } }},
+	} },
+	{ formats::SGBRG12_MTK, {
+		.name = "SGBRG12_MTK",
+		.format = formats::SGBRG12_MTK,
+		.v4l2Formats = { V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SGBRG12), },
+		.bitsPerPixel = 12,
+		.colourEncoding = PixelFormatInfo::ColourEncodingRAW,
+		.packed = true,
+		.pixelsPerGroup = 2,
+		.planes = {{ { 3, 1 }, { 0, 0 }, { 0, 0 } }},
+	} },
+	{ formats::SGRBG12_MTK, {
+		.name = "SGRBG12_MTK",
+		.format = formats::SGRBG12_MTK,
+		.v4l2Formats = { V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SGRBG12), },
+		.bitsPerPixel = 12,
+		.colourEncoding = PixelFormatInfo::ColourEncodingRAW,
+		.packed = true,
+		.pixelsPerGroup = 2,
+		.planes = {{ { 3, 1 }, { 0, 0 }, { 0, 0 } }},
+	} },
+	{ formats::SRGGB12_MTK, {
+		.name = "SRGGB12_MTK",
+		.format = formats::SRGGB12_MTK,
+		.v4l2Formats = { V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SRGGB12), },
+		.bitsPerPixel = 12,
+		.colourEncoding = PixelFormatInfo::ColourEncodingRAW,
+		.packed = true,
+		.pixelsPerGroup = 2,
+		.planes = {{ { 3, 1 }, { 0, 0 }, { 0, 0 } }},
+	} },
 
 	/* Compressed formats. */
 	{ formats::MJPEG, {
diff --git a/src/libcamera/formats.yaml b/src/libcamera/formats.yaml
index 539ac0b3..433c8251 100644
--- a/src/libcamera/formats.yaml
+++ b/src/libcamera/formats.yaml
@@ -183,4 +183,30 @@ formats:
   - SBGGR10_IPU3:
       fourcc: DRM_FORMAT_SBGGR10
       mod: IPU3_FORMAT_MOD_PACKED
+
+  - SRGGB10_MTK:
+      fourcc: DRM_FORMAT_SRGGB10
+      mod: MEDIATEK_FORMAT_MOD_BAYER_PACKED
+  - SGRBG10_MTK:
+      fourcc: DRM_FORMAT_SGRBG10
+      mod: MEDIATEK_FORMAT_MOD_BAYER_PACKED
+  - SGBRG10_MTK:
+      fourcc: DRM_FORMAT_SGBRG10
+      mod: MEDIATEK_FORMAT_MOD_BAYER_PACKED
+  - SBGGR10_MTK:
+      fourcc: DRM_FORMAT_SBGGR10
+      mod: MEDIATEK_FORMAT_MOD_BAYER_PACKED
+
+  - SRGGB12_MTK:
+      fourcc: DRM_FORMAT_SRGGB12
+      mod: MEDIATEK_FORMAT_MOD_BAYER_PACKED
+  - SGRBG12_MTK:
+      fourcc: DRM_FORMAT_SGRBG12
+      mod: MEDIATEK_FORMAT_MOD_BAYER_PACKED
+  - SGBRG12_MTK:
+      fourcc: DRM_FORMAT_SGBRG12
+      mod: MEDIATEK_FORMAT_MOD_BAYER_PACKED
+  - SBGGR12_MTK:
+      fourcc: DRM_FORMAT_SBGGR12
+      mod: MEDIATEK_FORMAT_MOD_BAYER_PACKED
 ...
diff --git a/src/libcamera/v4l2_pixelformat.cpp b/src/libcamera/v4l2_pixelformat.cpp
index 5551c62e..8c6c818a 100644
--- a/src/libcamera/v4l2_pixelformat.cpp
+++ b/src/libcamera/v4l2_pixelformat.cpp
@@ -161,6 +161,14 @@ const std::map<V4L2PixelFormat, V4L2PixelFormat::Info> vpf2pf{
 		{ formats::SGRBG10_CSI2P, "10-bit Bayer GRGR/BGBG Packed" } },
 	{ V4L2PixelFormat(V4L2_PIX_FMT_SRGGB10P),
 		{ formats::SRGGB10_CSI2P, "10-bit Bayer RGRG/GBGB Packed" } },
+	{ V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SBGGR10),
+		{ formats::SBGGR10_MTK, "10-bit Bayer BGGR MTISP Packed" } },
+	{ V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SGBRG10),
+		{ formats::SGBRG10_MTK, "10-bit Bayer GBRG MTISP Packed" } },
+	{ V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SGRBG10),
+		{ formats::SGRBG10_MTK, "10-bit Bayer GRBG MTISP Packed" } },
+	{ V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SRGGB10),
+		{ formats::SRGGB10_MTK, "10-bit Bayer RGGB MTISP Packed" } },
 	{ V4L2PixelFormat(V4L2_PIX_FMT_SBGGR12),
 		{ formats::SBGGR12, "12-bit Bayer BGBG/GRGR" } },
 	{ V4L2PixelFormat(V4L2_PIX_FMT_SGBRG12),
@@ -177,6 +185,14 @@ const std::map<V4L2PixelFormat, V4L2PixelFormat::Info> vpf2pf{
 		{ formats::SGRBG12_CSI2P, "12-bit Bayer GRGR/BGBG Packed" } },
 	{ V4L2PixelFormat(V4L2_PIX_FMT_SRGGB12P),
 		{ formats::SRGGB12_CSI2P, "12-bit Bayer RGRG/GBGB Packed" } },
+	{ V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SBGGR12),
+		{ formats::SBGGR12_MTK, "12-bit Bayer BGGR MTISP Packed" } },
+	{ V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SGBRG12),
+		{ formats::SGBRG12_MTK, "12-bit Bayer GBRG MTISP Packed" } },
+	{ V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SGRBG12),
+		{ formats::SGRBG12_MTK, "12-bit Bayer GRBG MTISP Packed" } },
+	{ V4L2PixelFormat(V4L2_PIX_FMT_MTISP_SRGGB12),
+		{ formats::SRGGB12_MTK, "12-bit Bayer RGGB MTISP Packed" } },
 	{ V4L2PixelFormat(V4L2_PIX_FMT_SBGGR14),
 		{ formats::SBGGR14, "14-bit Bayer BGBG/GRGR" } },
 	{ V4L2PixelFormat(V4L2_PIX_FMT_SGBRG14),
-- 
2.25.1

